strawberryfields/configuration.py

```diff
@@ -15,138 +15,229 @@
 This module contains the :class:`Configuration` class, which is used to
 load, store, save, and modify configuration options for Strawberry Fields.
 """
-import os
 import logging as log
+import os
 
 import toml
 from appdirs import user_config_dir
 
 log.getLogger()
 
 
-DEFAULT_CONFIG = {
-    "api": {
-        "authentication_token": "",
-        "hostname": "localhost",
-        "use_ssl": True,
-        "port": 443,
-        "debug": False}
-}
+class ConfigurationError(Exception):
+    """Exception used for configuration errors"""
+
+
+def load_config(filename="config.toml", **kwargs):
+    """Load configuration from keyword arguments, configuration file or
+    environment variables.
 
-BOOLEAN_KEYS = ("debug", "use_ssl")
+    .. note::
+
+        The configuration object (a nested dictionary) will be created based
+        on the following (order defines the importance, going from most
+        important to least important):
+
+        1. keyword arguments passed to ``load_config``
+        2. data contained in environmental variables (if any)
+        3. data contained in a configuration file (if exists)
+
+    Keyword arguments:
+        filename (str): the name of the configuration file to look for
+        authentication_token (str): the token to be used for user
+            authentication
+        hostname (str): the name of the host to connect to
+        use_ssl (bool): specifies if requests should be sent using SSL
+        port (int): the port to be used when connecting to the remote service
+        debug (bool): determines if the debugging mode is requested
+
+    Returns:
+        dict[str, dict[str, Union[str, bool, int]]]: the configuration
+            object
+    """
+    config = create_config_object()
+
+    parsed_config, _ = look_for_config_in_file(filename=filename)
+
+    if parsed_config is not None:
+        update_with_other_config(config, other_config=parsed_config)
+    else:
+        log.info("No Strawberry Fields configuration file found.")
 
+    update_from_environment_variables(config)
+
+    config_from_keyword_arguments = {"api": kwargs}
```

Comment:
It might make sense to define the kwargs as 'belonging' to the "api" section when the defaults/types are defined.

Reply:
Would that be referring to the defaults used in `create_config`?

One of the problems I came across was that if I use `create_config` and `kwargs` here, then having defaults will result in an unexpected overwrite.

That is
```python
In [1]: import strawberryfields.configuration as conf
In [3]:     config = { 
   ...:         "api": { 
   ...:             "authentication_token": 'NotDefaultAuth', 
   ...:             "hostname": 'NotDefaultHostname', 
   ...:             "use_ssl": True, 
   ...:             "port": 123456789, 
   ...:             "debug": False 
   ...:             } 
   ...:     } 
   ...: other_config = conf.create_config(authentication_token="MyToken") 
   ...: conf.update_config(config, other_config) 
   ...: config                                                                  
Out[3]: 
{'api': {'authentication_token': 'MyToken',
  'hostname': 'localhost',
  'use_ssl': True,
  'port': 443,
  'debug': False}}
```
where  `'hostname'` and `'port'` were overwritten, whereas they were not defined in `kwargs`.

Reply:
Ah, I see. This is because the `other_config` is replacing the default values 🤔